File: Python/20.Valid Parentheses.py
#看字符串中括号是否匹配
class Solution(object):
    def isValid(self, s):
        l =len(s)
        stack = []
        for ch in s:
            if ch in ['(','{','[']:
                stack.append(ch)
            elif ch in [')',']','}']:
                if len(stack)==0:
                    return False
                if ch==')' and stack[-1]=='(':
                    stack.pop()
                elif ch==']' and stack[-1]=='[':
                    stack.pop()
                elif ch=='}' and stack[-1]=='{':
                    stack.pop()
                else:
                    return False
            else:
                return False
        if len(stack)==0:
            return True
        else:
            return False


s = Solution()
print(s.isValid('{([()]())}'))

File: C++/String/20.Valid Parentheses.cpp
/*
给出一组大中小括号，判断它们是否成对。
*/
/*
思路：用堆栈。
*/
#include <iostream>
#include <string>
#include <stack>
using namespace std;

class Solution {
public:
	bool isValid(string s) {
		stack<char> mystack;
        for(char ch:s)
        {
        	switch(ch)
        	{
        		case '{':
        		case '[':
        		case '(': mystack.push(ch);break;
        		case "}": if(!mystack.empty() && mystack.top()=='{') mystack.pop();else return false; break;
        		case "]": if(!mystack.empty() && mystack.top()=='[') mystack.pop();else return false; break;
        		case ")": if(!mystack.empty() && mystack.top()=='(') mystack.pop();else return false; break;
        		default:	;
        	}
        }
        return mystack.empty();
    }
};

int main()
{
	Solution s;;
	cout<<s.isValid("[]{})")<<endl;
}

